### ISSUE TITLE
Weighted WENO stencils for stretched grids

### ISSUE THREAD

**Issue by glwagner** [CliMA/Oceananigans.jl#1704]
In [A simple algorithm to improve the performance of the WENO scheme on non-uniform grids](https://link.springer.com/article/10.1007/s10409-017-0715-2) a simple method for generalizing WENO stencils for stretched grids is described. This scheme is easy to implement and limits to what we already have implemented in the case of a regular grid.

**Comment by glwagner** [comment]
To implement this scheme we first need to reformulate all the stencils in https://github.com/CliMA/Oceananigans.jl/blob/master/src/Advection/weno_fifth_order.jl in terms of interpolation operators.

The next step is to use weighted interpolation operators on stretched grids according to this definition:

![image](https://user-images.githubusercontent.com/15271942/119732042-2a237900-be24-11eb-93e4-4c2f3fd053a6.png)

"That's it."

**Comment by tomchor** [comment]
Any clue as to what amount of error we're introducing by using WENO5+`StretchedGrids`? Maybe enough to put an approx upper bound to the level of stretching we should use? I'm assuming this only becomes important when the grid stretching is too aggressive but who knows...

Also I wonder how this plays out with https://github.com/CliMA/Oceananigans.jl/issues/1705

**Comment by glwagner** [comment]
No clue, but the reference tests those cases and includes a discussion.

**Comment by francispoulin** [comment]
If we wanted to quantify the difference, we can simply find an exact solution and simulate it and compute the error.  There are lots of examples to choose from.

**Comment by glwagner** [comment]
Here's a couple results picked from the paper:

# Stationary vortex on a grid with "random" spacing

![image](https://user-images.githubusercontent.com/15271942/124931772-c57d4180-dfbf-11eb-9eec-647391df0bec.png)

# Stationary vortex on a curvilinear distorted grid

![image](https://user-images.githubusercontent.com/15271942/124931931-e3e33d00-dfbf-11eb-821a-b537a99586dd.png)

Contour plots of a perfectly circular vortex are good at highlighting the differences. Their "production" case of flow around an airfoil is more subtle:

![image](https://user-images.githubusercontent.com/15271942/124932064-feb5b180-dfbf-11eb-9963-8343fb5f0c54.png)

**Comment by glwagner** [comment]
In the above, WENO-JS is the scheme we have now, and WENO-NM is the "weighted" scheme they propose.

The weighted scheme is extremely simple. It's just a little algebra to get it to work for us.

**Comment by francispoulin** [comment]
Thanks @glwagner ,  Certainly Figure 5 look sdifferent and that is good inspiration to consider how to impliment WENO-NM.

**Comment by tomchor** [comment]
@glwagner  I've been thinking about the WENO5 algorithm based on our discussion in https://github.com/CliMA/Oceananigans.jl/discussions/2054, and I just wanted to ask a quick question to make sure I understand the problem in this issue.

If I understand correctly, the WENO5 scheme calculates the advection with different advection schemes (all of which are correct regardless of the grid) and then does a weighted sum of these solutions according so some criterion. So the issue with WENO5 and StretchedGrids is not that the advection calculated is "wrong", it's just that the weighting of the solutions is done in a non-optimal manner, and therefore introduces errors that could be avoided, right?

Basically I'm asking the question: is it the case the shouldn't use WENO5 with stretched grids because the answer is wrong, period? Or we can use it, it's just not gonna be as accurate as it potentially could be?

**Comment by glwagner** [comment]
The results in [A simple algorithm to improve the performance of the WENO scheme on non-uniform grids](https://link.springer.com/article/10.1007/s10409-017-0715-2) show that WENO reverts to 2nd-order accuracy on stretched grids.

However, for any given problem, the convergence rate is only one aspect of the accuracy of a solution. It's possible that WENO5 is still more accurate than any other numerical scheme, even if the solution only _converges_ to the true solution at a 2nd order rate.
